backends-common/elasticsearch-v8/src/main/java/org/apache/james/backends/es/v8/search/ScrolledSearch.java

```diff
@@ -0,0 +1,109 @@
+import java.util.function.Consumer;
+
+import org.apache.james.backends.es.v8.ReactorElasticSearchClient;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.github.fge.lambdas.Throwing;
+
+import co.elastic.clients.elasticsearch._types.Time;
+import co.elastic.clients.elasticsearch.core.ClearScrollRequest;
+import co.elastic.clients.elasticsearch.core.ScrollRequest;
+import co.elastic.clients.elasticsearch.core.ScrollResponse;
+import co.elastic.clients.elasticsearch.core.SearchRequest;
+import co.elastic.clients.elasticsearch.core.search.Hit;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+
+public class ScrolledSearch {
+    private static final Time TIMEOUT = new Time.Builder()
+        .time("1 minute")
+        .build();
+
+    private final ReactorElasticSearchClient client;
+    private final SearchRequest searchRequest;
+
+    public ScrolledSearch(ReactorElasticSearchClient client, SearchRequest searchRequest) {
+        this.client = client;
+        this.searchRequest = searchRequest;
+    }
+
+    public <T> Flux<Hit<T>> searchHits() {
+        return searchResponses()
+            .concatMap(searchResponse -> Flux.fromIterable(searchResponse.hits().hits()));
+    }
+
+    private Flux<ScrollResponse> searchResponses() {
+        return Flux.push(sink -> {
+            AtomicReference<Optional<String>> scrollId = new AtomicReference<>(Optional.empty());
+            sink.onRequest(numberOfRequestedElements -> next(sink, scrollId, numberOfRequestedElements));
+
+            sink.onDispose(() -> close(scrollId));
+        });
+    }
+
+    private void next(FluxSink<ScrollResponse> sink, AtomicReference<Optional<String>> scrollId, long numberOfRequestedElements) {
+        if (numberOfRequestedElements <= 0) {
+            return;
+        }
+
+        Consumer<ScrollResponse> onResponse = searchResponse -> {
+            scrollId.set(Optional.of(searchResponse.scrollId()));
+            sink.next(searchResponse);
+
+            if (searchResponse.hits().hits().isEmpty()) {
+                sink.complete();
+            } else {
+                next(sink, scrollId, numberOfRequestedElements - 1);
+            }
+        };
+
+        Consumer<Throwable> onFailure = sink::error;
+
+        buildRequest(scrollId.get())
+            .subscribe(onResponse, onFailure);
+    }
+
+    private Mono<ScrollResponse<ObjectNode>> buildRequest(Optional<String> scrollId) {
+        return scrollId.map(id ->
+            client.scroll(new ScrollRequest.Builder()
+                    .scrollId(scrollId.get())
+                    .scroll(TIMEOUT)
+                    .build()))
+            //.orElseGet(() -> client.search(searchRequest));
+            .get(); // TODO: scroll and search return different responses now, thus this trick not working...
+                    // not sure of the best way to solve that yet
```

Comment:
Likely the level of indirection needs to be escalated...

Reply:
Tried quickly without success... will dig into it after I manage to make the rest work